Source/Editor/DualityEditor/Forms/ObjectRefSelectionDialog.cs

```diff
@@ -0,0 +1,281 @@
+﻿using System;
+using System.Drawing;
+using System.Linq;
+using System.Windows.Forms;
+
+using Aga.Controls.Tree;
+using Aga.Controls.Tree.NodeControls;
+using Duality.Resources;
+
+namespace Duality.Editor.Forms
+{
+	public partial class ObjectRefSelectionDialog : Form
+	{
+		public class ReferenceNode : Node
+		{
+			public ReferenceNode(IContentRef resource) : base(resource.Name)
+			{
+				this.Name = resource.Name;
+				this.Path = resource.FullName;
+
+				this.ResourceReference = resource;
+			}
+
+			public ReferenceNode(GameObject gameObject) : base(gameObject.Name)
+			{
+				this.Name = gameObject.Name;
+				this.Path = gameObject.FullName;
+
+				this.GameObjectReference = gameObject;
+			}
+
+			public ReferenceNode(Component component) : base(component.GetType().GetTypeCSCodeName())
+			{
+				this.Name = component.GameObj.FullName;
+				this.Path = component.GetType().GetTypeCSCodeName();
+
+				this.ComponentReference = component;
+			}
+

```

Comment:
Code Style: 
* Fields should be next to fields, properties should be next to properties and methods should be next to methods. Do not mix them if it can be avoided within reason. 
* Keep statics at the top, instance methods below. 
* If the number of members of the same type grows beyond what can be easily observed with a full overview, try to group semantically where it makes sense. 
* Use two newlines to separate different member types, one newline to separate semantic groups within the same type, no newline between items of the same semantic group and type.

Example:
```csharp
public class Foo
{
	private int field1;
	private int field2;
	private int field3;
	
	
	public int Property1
	{
		get { /* ... */ }
		set { /* ... */ }
	}
	public int Property2
	{
		get { /* ... */ }
		set { /* ... */ }
	}
	public int Property3
	{
		get { /* ... */ }
		set { /* ... */ }
	}
	
	
	public Foo()
	{
		// ...
	}
	
	public void Method1()
	{
		// ...
	}
	public void Method2()
	{
		// ...
	}
	public void Method3()
	{
		// ...
	}
}
```